File: Chapter 08/Python/section_8_2_1_declarative_querying/demo2.py
# demo.py
from sqlalchemy import create_engine, text
from sqlalchemy.orm import sessionmaker
from .models import Base, User

class Demo2:
    """
    THE SCENARIO REPOSITORY:
    Demonstrates the architectural shift from Imperative (Raw SQL) 
    to Declarative (ORM) querying.
    """

    @staticmethod
    def run_query_comparison() -> None:
        print("\n=== Section 8.2.1: Declarative Querying (Raw SQL vs ORM) ===")
        print("SCENARIO: The database contains 4 users. We need to find all active users over age 21, sorted alphabetically.")

        # 1. Setup the in-memory database and ORM Session
        engine = create_engine('sqlite:///:memory:', echo=False)
        Base.metadata.create_all(engine) # The ORM automatically builds the tables!
        Session = sessionmaker(bind=engine)
        session = Session()

        # 2. Seed the database with test data
        session.add_all([
            User(first_name="Alice", last_name="Smith", age=25, is_active=True),
            User(first_name="Bob", last_name="Jones", age=19, is_active=True),      # Fails: Too young
            User(first_name="Charlie", last_name="Brown", age=30, is_active=False), # Fails: Inactive
            User(first_name="Diana", last_name="Prince", age=28, is_active=True)
        ])
        session.commit()
        print("SETUP: 4 Users inserted into the database.\\n")


        # --- THE OLD WAY (IMPERATIVE) ---
        print("--- 1. The Old Way (Imperative / Raw SQL) ---")
        raw_sql = "SELECT * FROM users WHERE age > 21 AND is_active = 1 ORDER BY last_name;"
        print(f"  [Action] Executing Raw String: {raw_sql}")
        
        # We use SQLAlchemy's text() wrapper just to execute the raw string safely
        raw_result = session.execute(text(raw_sql))
        
        found_raw = []
        for row in raw_result:
            # We have to access rows by exact column string names, which is brittle
            found_raw.append(f"{row.first_name} {row.last_name}")
            
        print(f"  [Result] Found: [{', '.join(found_raw)}]")
        print("  [Lesson] The burden is on you. If you mistyped 'is_active' as 'isActive' inside that string,")
        print("           your code would compile perfectly, but crash in production.\\n")


        # --- THE MODERN WAY (DECLARATIVE) ---
        print("--- 2. The Modern Way (Declarative / ORM) ---")
        print("  [Action] Building a query object using native Python syntax...")
        
        # This is Listing 8.5 from the textbook!
        users = session.query(User).filter(User.age > 21, User.is_active == True).order_by(User.last_name).all()

        found_orm = []
        for u in users:
            # We access data using strongly-typed object properties
            found_orm.append(f"{u.first_name} {u.last_name}")

        print(f"  [Result] Found: [{', '.join(found_orm)}]")
        print("  [Lesson] The ORM translates your Python code into SQL safely behind the scenes.")
        print("           If you rename 'User.age' to 'User.years_old', your IDE will instantly flag the error.")

        session.close()

File: Chapter 08/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter08/section_8_2_1_declarative_querying/Demo2.java
// src/main/java/com/grokking/section2_declarativequerying/Demo2.java
package com.grokkingsoftwarearchitecture.chapter08.section_8_2_1_declarative_querying;

import org.hibernate.Session;
import java.util.List;

public class Demo2 {
    public static void runQueryComparison() {
        System.out.println("\n=== Section 8.2.1: Declarative Querying (Raw SQL vs ORM) ===");
        System.out.println("SCENARIO: The database contains 4 users. We need to find all active users over age 21, sorted alphabetically.");

        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            session.beginTransaction();

            // 1. Seed the database with test data
            session.persist(new User("Alice", "Smith", 25, true));
            session.persist(new User("Bob", "Jones", 19, true));      // Fails: Too young
            session.persist(new User("Charlie", "Brown", 30, false)); // Fails: Inactive
            session.persist(new User("Diana", "Prince", 28, true));
            
            session.getTransaction().commit();
            System.out.println("SETUP: 4 Users inserted into the database.\n");

            
            // --- THE OLD WAY (IMPERATIVE) ---
            System.out.println("--- 1. The Old Way (Imperative / Raw SQL) ---");
            String rawSql = "SELECT * FROM users WHERE age > 21 AND is_active = 1 ORDER BY last_name";
            System.out.println("  [Action] Executing Raw String: " + rawSql);
            
            // Forcing Hibernate to execute a raw SQL string
            List<User> rawUsers = session.createNativeQuery(rawSql, User.class).list();
            
            System.out.print("  [Result] Found: [");
            for (int i = 0; i < rawUsers.size(); i++) {
                System.out.print(rawUsers.get(i).getFirstName() + " " + rawUsers.get(i).getLastName());
                if (i < rawUsers.size() - 1) System.out.print(", ");
            }
            System.out.println("]");
            System.out.println("  [Lesson] The burden is on you. If you mistyped 'is_active' inside that string,");
            System.out.println("           your code would compile perfectly, but crash in production.\n");


            // --- THE MODERN WAY (DECLARATIVE) ---
            System.out.println("--- 2. The Modern Way (Declarative / ORM) ---");
            System.out.println("  [Action] Building a query object using native Hibernate Query Language (HQL)...");
            
            // Declarative approach using HQL (Hibernate Query Language)
            String hql = "FROM User u WHERE u.age > 21 AND u.isActive = true ORDER BY u.lastName";
            List<User> ormUsers = session.createQuery(hql, User.class).list();

            System.out.print("  [Result] Found: [");
            for (int i = 0; i < ormUsers.size(); i++) {
                System.out.print(ormUsers.get(i).getFirstName() + " " + ormUsers.get(i).getLastName());
                if (i < ormUsers.size() - 1) System.out.print(", ");
            }
            System.out.println("]");
            System.out.println("  [Lesson] The ORM translates your HQL into safe SQL behind the scenes.");
            System.out.println("           It maps the results directly into strongly-typed Java objects.");
        }
    }
}